core/trino-main/src/main/java/io/trino/exchange/SpoolingExchangeDataSource.java

```diff
@@ -0,0 +1,142 @@
+    // It is modified (assigned to null) when the ExchangeOperator is closed.
+    // It doesn't have to be declared as volatile as the nullification of this variable doesn't have to be immediately visible to other threads.
+    // However since close can be called at any moment this variable has to be accessed in a safe way (avoiding "check-then-use").
+    private ExchangeSource exchangeSource;
+    private final List<ExchangeSourceHandle> exchangeSourceHandles;
+    private final LocalMemoryContext systemMemoryContext;
+    private volatile boolean closed;
+
+    public SpoolingExchangeDataSource(
+            ExchangeSource exchangeSource,
+            List<ExchangeSourceHandle> exchangeSourceHandles,
+            LocalMemoryContext systemMemoryContext)
+    {
+        // this assignment is expected to be followed by an assignment of a final field to ensure safe publication
+        this.exchangeSource = requireNonNull(exchangeSource, "exchangeSource is null");
+        this.exchangeSourceHandles = ImmutableList.copyOf(requireNonNull(exchangeSourceHandles, "exchangeSourceHandles is null"));
+        this.systemMemoryContext = requireNonNull(systemMemoryContext, "systemMemoryContext is null");
+    }
+
+    @Override
+    public Slice pollPage()
+    {
+        ExchangeSource exchangeSource = this.exchangeSource;
+        if (exchangeSource == null) {
+            return null;
+        }
+
+        Slice data = exchangeSource.read();
+        systemMemoryContext.setBytes(exchangeSource.getMemoryUsage());
+
+        // If the data source has been closed in a meantime reset memory usage back to 0
+        if (closed) {
+            systemMemoryContext.setBytes(0);
+        }
+
+        return data;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        ExchangeSource exchangeSource = this.exchangeSource;
+        if (exchangeSource == null) {
+            return true;
+        }
+        return exchangeSource.isFinished();
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        ExchangeSource exchangeSource = this.exchangeSource;
+        if (exchangeSource == null) {
+            return immediateVoidFuture();
+        }
+        return toListenableFuture(exchangeSource.isBlocked());
+    }
+
+    @Override
+    public void addInput(ExchangeInput input)
+    {
+        SpoolingExchangeInput exchangeInput = (SpoolingExchangeInput) input;
+        // Only a single input is expected when the spooling exchange is used.
+        // The engine adds the same input to every instance of the ExchangeOperator.
+        // Since the ExchangeDataSource is shared between ExchangeOperator instances
+        // the same input may be delivered multiple times.
+        checkState(
+                exchangeInput.getExchangeSourceHandles().equals(exchangeSourceHandles),
+                "split is expected to contain an identical exchangeSourceHandles list: %s != %s",
```

Comment:
preexisting - but the error message is no longer valid as we are not operating on splits now. Please update in separate commit.